File: tests/javascript/UserRepository.js
/**
 * Repository for managing User data persistence.
 */
class UserRepository {
  constructor() {
    this.database = new Map();
    this.idCounter = 1;
  }

  /**
   * Save a user to the database.
   * If the user has an ID, it updates; otherwise creates a new user.
   * @param {User} user - The user to save
   * @returns {User} The saved user
   */
  save(user) {
    if (user.id === null) {
      user.id = this.idCounter++;
    }
    this.database.set(user.id, user);
    return user;
  }

  /**
   * Find a user by their ID.
   * @param {number} id - The user ID
   * @returns {User|null} The user or null if not found
   */
  findById(id) {
    return this.database.get(id) || null;
  }

  /**
   * Find a user by their username.
   * @param {string} username - The username to search for
   * @returns {User|null} The user or null if not found
   */
  findByUsername(username) {
    for (const user of this.database.values()) {
      if (user.username === username) {
        return user;
      }
    }
    return null;
  }

  /**
   * Find all active users.
   * @returns {User[]} Array of active users
   */
  findAllActive() {
    return Array.from(this.database.values()).filter(user => user.active);
  }

  /**
   * Delete a user by ID.
   * @param {number} id - The user ID to delete
   * @returns {boolean} True if deleted, false otherwise
   */
  deleteById(id) {
    return this.database.delete(id);
  }

  /**
   * Get count of all users.
   * @returns {number} Total number of users
   */
  count() {
    return this.database.size;
  }
}

module.exports = UserRepository;

File: tests/go/user_repository.go
package main

import "sync"

// UserRepository manages User data persistence.
// Uses in-memory storage for demonstration purposes.
type UserRepository struct {
	database  map[int64]*User
	idCounter int64
	mu        sync.RWMutex
}

// NewUserRepository creates a new UserRepository instance.
func NewUserRepository() *UserRepository {
	return &UserRepository{
		database:  make(map[int64]*User),
		idCounter: 1,
	}
}

// Save saves a user to the database.
// If the user has an ID, it updates; otherwise creates a new user.
func (r *UserRepository) Save(user *User) *User {
	r.mu.Lock()
	defer r.mu.Unlock()

	if user.ID == nil {
		id := r.idCounter
		user.ID = &id
		r.idCounter++
	}

	r.database[*user.ID] = user
	return user
}

// FindByID finds a user by their ID.
func (r *UserRepository) FindByID(id int64) *User {
	r.mu.RLock()
	defer r.mu.RUnlock()

	return r.database[id]
}

// FindByUsername finds a user by their username.
func (r *UserRepository) FindByUsername(username string) *User {
	r.mu.RLock()
	defer r.mu.RUnlock()

	for _, user := range r.database {
		if user.Username == username {
			return user
		}
	}
	return nil
}

// FindAllActive finds all active users.
func (r *UserRepository) FindAllActive() []*User {
	r.mu.RLock()
	defer r.mu.RUnlock()

	var activeUsers []*User
	for _, user := range r.database {
		if user.Active {
			activeUsers = append(activeUsers, user)
		}
	}
	return activeUsers
}

// DeleteByID deletes a user by ID.
func (r *UserRepository) DeleteByID(id int64) bool {
	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.database[id]; exists {
		delete(r.database, id)
		return true
	}
	return false
}

// Count returns the total number of users.
func (r *UserRepository) Count() int {
	r.mu.RLock()
	defer r.mu.RUnlock()

	return len(r.database)
}
